service-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioningEnabledDeviceBackend.java

```diff
@@ -0,0 +1,90 @@
+
+import java.net.HttpURLConnection;
+import java.security.cert.X509Certificate;
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import javax.security.auth.x500.X500Principal;
+
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.credentials.X509CertificateCredential;
+import org.eclipse.hono.service.management.credentials.X509CertificateSecret;
+
+import io.opentracing.Span;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+
+/**
+ * Interface that adds Auto-Provisioning to the DeviceBackend.
+ */
+public interface AutoProvisioningEnabledDeviceBackend extends DeviceBackend, AutoProvisioningService {
+
+    @Override
+    default Future<OperationResult<String>> provisionDevice(
+            final String tenantId,
+            final X509Certificate clientCertificate,
+            final Span span) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(clientCertificate);
+        Objects.requireNonNull(span);
+
+        span.log("Start auto-provisioning");
+        final String comment = "Auto-provisioned at " + Instant.now().toString();
+
+        // 1. create device
+        final Device device = new Device().setEnabled(true).putExtension("comment", comment);
+        final Promise<OperationResult<Id>> createDeviceAttempt = Promise.promise();
+        createDevice(tenantId, Optional.empty(), device, span, createDeviceAttempt);
+
+        return createDeviceAttempt.future()
+                .compose(r -> {
+                    if (r.isError()) {
+                        return Future.succeededFuture(OperationResult.ok(r.getStatus(),
+                                "Auto-provisioning failed: device could not be created", Optional.empty(),
+                                Optional.empty()));
+                    }
+
+                    // 2. set the certificate credential
+                    final X509CertificateCredential certCredential = new X509CertificateCredential()
+                            .setSecrets(List.of(new X509CertificateSecret()));
+                    certCredential.setEnabled(true).setComment(comment)
+                            .setAuthId(clientCertificate.getSubjectX500Principal().getName(X500Principal.RFC2253));
+
+                    final String deviceId = r.getPayload().getId();
+
+                    final Promise<OperationResult<Void>> credPromise = Promise.promise();
+                    set(tenantId, deviceId, Optional.empty(), List.of(certCredential), span, credPromise);
+                    return credPromise.future()
+                            .map(v -> {
+                                if (v.isError()) {
+                                    return OperationResult.ok(v.getStatus(),
+                                            "Auto-provisioning failed: credentials could not be set for device ["
```

Comment:
shouldn't the creation of the device identity be reverted in this case? Otherwise, we would end up with a *half-provisioned* device, wouldn't we?